core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceConcurrencyTest.java

```diff
@@ -0,0 +1,311 @@
+    assertThat(d.get(2)).isNotNull();
+    assertThat(supplier.counter.get()).isEqualTo(1);
+
+    for (int i = 0; i < 10; i++) {
+      // Call forceRefresh simultaneously
+      ListenableFuture<List<Object>> all =
+          Futures.allAsList(
+              executor.submit(instance::forceRefresh),
+              executor.submit(instance::forceRefresh),
+              executor.submit(instance::forceRefresh));
+      try {
+        all.get();
+      } catch (Exception e) {
+      }
+
+      ListenableFuture<List<Object>> allData2 =
+          Futures.allAsList(
+              executor.submit(instance::getSslData),
+              executor.submit(instance::getSslData),
+              executor.submit(instance::getSslData));
+      try {
+        allData2.get();
+      } catch (Exception e) {
+      }
+      assertThat(supplier.counter.get()).isEqualTo(2 + i);
+      Thread.sleep(100);
+    }
+  }
+
+  @Test
+  public void testCloudSqlInstanceCorrectlyRefreshesInstanceData() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    ListenableFuture<KeyPair> keyPairFuture =
+        Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    ListeningScheduledExecutorService executor = newTestExecutor();
+    TestDataSupplier supplier = new TestDataSupplier(false);
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "a:b:c",
+            supplier,
+            AuthType.PASSWORD,
+            new TestCredentialFactory(),
+            executor,
+            keyPairFuture,
+            newRateLimiter());
+
+    assertThat(supplier.counter.get()).isEqualTo(0);
+    Thread.sleep(100);
+    SslData data = instance.getSslData();
+    assertThat(data).isNotNull();
+  }
+
+  @Test
+  public void testRefreshWhenRefreshRequestAlwaysFails() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    ListenableFuture<KeyPair> keyPairFuture =
+        Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    ListeningScheduledExecutorService executor = newTestExecutor();
+
+    InstanceDataSupplier supplier =
+        (CloudSqlInstanceName instanceName,
+            AccessTokenSupplier accessTokenSupplier,
+            AuthType authType,
+            ListeningScheduledExecutorService exec,
+            ListenableFuture<KeyPair> keyPair) -> {
+          ListenableFuture<?> f =
+              exec.submit(
+                  () -> {
+                    throw new RuntimeException("always fails");
+                  });
+          f.get(); // this will throw an ExecutionException
+          return null;
+        };
+
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "a:b:c",
+            supplier,
+            AuthType.PASSWORD,
+            new TestCredentialFactory(),
+            executor,
+            keyPairFuture,
+            newRateLimiter());
+
+    Thread.sleep(500);
+    Assert.assertThrows(RuntimeException.class, () -> instance.getSslData());
```

Comment:
The key difference is that this one throws an exception from within a scheduled executor task. This used to exacerbate the deadlock problem.

Reply:
Is this behavior covered by `testCloudSqlInstanceRefreshesConsistentlyWithoutRaceConditions`? It looks like a duplicate test to my eyes.

Reply:
This test has a single CloudSqlInstance and ensures that it successfully refreshes the correct number of times. Thus proving that the balking behavior of forceRefresh() works right. 

`testCloudSqlInstanceRefreshesConsistentlyWithoutRaceConditions` tries to cause a race condition and deadlock with multiple CloudSqlInstances and many more threads. It is fundimentally a different test. 

While developing, if there was a bug in my code, one of these tests would fail and the other would pass. So I think they are both useful.

Reply:
Right now this test proves that the CloudSqlInstance calls the underlying InstanceDataSupplier, which is a variation of [this test](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java#L65). If you'd like to test that force refresh functions correctly, then I'd expect a call that fails, a force refresh, and then a call that succeeeds.